### PR TITLE
[FIX] mail: fix wrong escaping of html entities when create link

### LINKED ISSUE #70485: [14.0] mail: issue with hypertext formating
**Impacted versions: 14.0**

**Steps to reproduce:**
create a transaction link like this:
`https://exemple.com/website_payment/pay?reference=100516376&amount=300.0&currency_id=1&partner_id=75178&order_id=533121&company_id=1&access_token=65433c7c1d78a5cb391527eb4035b11b052aeb53b89e9e7526d0b77021d02020`

- create an order, click "send by email"
- paste link in the email
- send it
--> Issue the render of hyperlink is wrong:
During editting
image

Final email
image

`300.0&currency_id=1` is remplaced by `300.0¤cy_id=1`

If customer copy-paste the link, the copied link is wrong.

Video/Screenshot link (optional):

https://user-images.githubusercontent.com/16716992/117309226-8416cb80-ae82-11eb-8137-3e21296dbf14.mov

Support ticket number submitted via odoo.com/help (optional):

### PR DESCRIPTION
**Pull request by Gorash**
Issue: `node.data` is an unescape text content and used as an escaped
content just after (set `innerHTML`). The emoji will be search on
unescaped and escaped content to post a message.

opw-2525028

### CODE REVIEW

**Comment by seb-odoo** [comment]
Hello, just a notice this looks somewhat related to https://github.com/odoo/odoo/pull/70950

**Comment by Gorash** [comment]
@seb-odoo @nle-odoo is this good for you?

File: addons/mail/static/src/js/utils.js
Comment (seb-odoo) [inline addons/mail/static/src/js/utils.js:74]: I feel like this should be done inside `linkify`, after the link detection, when actually building the tag (both href and url can be wrapped)? But it doesn't solve the other issue I have (see below).

File: addons/mail/static/tests/mail_utils_tests.js
Comment (seb-odoo) [inline addons/mail/static/tests/mail_utils_tests.js:38]: name suggestion: ```add_link() result should have &amp; when the given url text contains &amp;```

File: addons/mail/static/src/models/composer/composer.js
Comment (seb-odoo) [inline addons/mail/static/src/models/composer/composer.js:831]: I don't understand this change, could you please explain it?
It looks unrelated to opw and also it looks like no test is covering it?

File: addons/mail/static/tests/mail_utils_tests.js
Comment (seb-odoo) [inline addons/mail/static/tests/mail_utils_tests.js:41]: I tried to see how it would go with ```<p>https://test.example.com/test?&currency_id</p>```
and the result is actually broken the same, with `¤cy_id` after the fix.

So the fix is not correct, and a test should be added to cover that case too.

**Review by seb-odoo — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/js/utils.js
Comment (nle-odoo) [reply addons/mail/static/src/js/utils.js:74]: yes I think like this:

<https://test.example.com/test?&currency_id=3>

will link on `https://test.example.com/test?&currency_id=3>` I think because > becomes `&gt;` which becomes part of the link

File: addons/mail/static/tests/mail_utils_tests.js
Comment (nle-odoo) [reply addons/mail/static/tests/mail_utils_tests.js:41]: i'm not sure this is an issue, `<p>https://test.example.com/test?&currency_id</p>` is equivalent to `<p>https://test.example.com/test?¤cy_id</p>` in HTML so your use case looks like expected behavior

File: addons/mail/static/src/models/composer/composer.js
Comment (Gorash) [reply addons/mail/static/src/models/composer/composer.js:831]: This is to do the emoji check with and without the escaping. Because otherwise in the links or in the document icons such as `:'(` have an escape and therefore are no longer displayed as an icon.

**Comment by Gorash** [comment]
@seb-odoo @nle-odoo Following your comments, can we merge this already? This already fixes the outcome, if other cases are to be added we can do it after.

File: addons/mail/static/src/js/utils.js
Comment (seb-odoo) [reply addons/mail/static/src/js/utils.js:74]: @Gorash to answer your latest message, this comment still needs to be resolved before merging I believe.
And also the security team needs to validate any change in these methods due to the sensitivity of them.

**Comment by nle-odoo** [comment]
hi, I've made a new PR in 13.0 and taken into account the https://github.com/odoo/odoo/pull/74255#discussion_r689530883 comment, I also added some test assertions for different use case.

**Pull request closed without merging**
